### ISSUE TITLE
findfirst  , findlast and findall differences in needles

### ISSUE THREAD

**Issue by oxinabox** [JuliaLang/julia#37350]
`findfirst`, `findlast` and `findall` each support a different set of needles when the haystack is an AbstractString.
```julia
findfirst(::Union{Regex, AbstractString, Function, AbstractChar}, ::AbstractString)
# Everything

findall(::Union{Regex, AbstractString, Function}, ::AbstractString)
# No AbstractChar

findlast(::Union{AbstractString, Function, AbstractChar}, ::AbstractString)
# No Regex
```

We should make them consistent

**Comment by nalimilan** [comment]
See also `findnext` and `findprev`, which are respectively consistent with `findfirst` and `findlast`.

I'm not sure it's possible to get the last match of a regex with PCRE. That question may not be well-defined, as one can also write a regex to extract the last match of an expression, and pass that to `findfirst`. Of course as a brute-force method we could have `findlast` return `last(findall(...))`, but that's not a good implementation.

**Comment by stevengj** [comment]
Yes, PCRE doesn't support reverse-order searching.  (You can simulate by reversing the string and searching with a reversed regex, but automatically constructing a reversed regex in general seems to be a hard problem.)

*Update*: see also #6276 and #42086

**Comment by ultradian** [comment]
It seems to me that when I want to do `findprev` or `findlast` using a regex needle for a String, I'd want to do a forward order PCRE search. Like wanting 'A5' then 'A4'  in `findprev(r"A\d", "teststring 3A distrActing A4 and A5 target center teststring 2A distrActing A7 and A8 target", 46), where 46 is the center of the haystack.

Would a PR be appropriate for this?

**Comment by stevengj** [comment]
> It seems to me that when I want to do findprev or findlast using a regex needle for a String, I'd want to do a forward order PCRE search.

This would require you to search the entire string every time, whereas a reverse-order search can start from the end and stop as soon as a match is found.

**Comment by stevengj** [comment]
> `findall(::Union{Regex, AbstractString, Function}, ::AbstractString) # No AbstractChar`

(Note, by the way, that the lack of `AbstractChar` in `findall` has been fixed.)

**Comment by ultradian** [comment]
Oh duh. You're right. The links to the issues in your _Update_ above are very helpful in getting me to think more carefully.
